File: Java/src/ThreeSum.java
import java.util.*;

// Time complexity: O(n^2)
// Space complexity: O(1)
public class ThreeSum {
    public static void main(String[] args) {
        int[] arr = { 7, -6, 3, 8, -1, 8, -11 };
        int target = 0; // sum of three number in the above array should be equals target
        solution(arr, target, arr.length);
    }

    public static void solution(int[] a, int target, int n) {
        Arrays.sort(a);
        for (int i = 0; i < n; i++) {
            if (i == 0 || (a[i] != a[i - 1])) {
                int j = i + 1, k = n - 1;
                int tar = target - a[i];
                while (j < k) {
                    int twoSum = a[j] + a[k];
                    if (twoSum == tar) {
                        System.out.println(a[i] + " " + a[j] + " " + a[k]);
                        while (j < k && a[j] == a[j + 1])
                            j++;
                        while (j < k && a[k] == a[k - 1])
                            k--;
                        j++;
                        k--;
                    } else if (twoSum < tar) {
                        j++;
                    } else {
                        k--;
                    }
                }
            }
        }
    }
}

File: Python/Arrays/ThreeSum.py
# Time complexity: O(n^2)
# Space complexity: O(1)

def ThreeSum():
    arr = [7, -6, 3, 8, -1, 8, -11]
    target = 0
    Solution(arr, target, len(arr))


def Solution(a, target, n):
    a.sort()
    for i in range(n):
        if (i == 0 or a[i] != a[i-1]):
            j = i+1
            k = n - 1
            tar = target - a[i]
            while (j < k):
                two_sum = a[j] + a[k]
                if (two_sum == tar):
                    print("{} {} {}".format(a[i], a[j], a[k]))
                    while (j < k and a[j] == a[j+1]):
                        j += 1
                    while (j < k and a[k] == a[k-1]):
                        k -= 1
                    j += 1
                    k -= 1
                elif (two_sum < tar):
                    j += 1
                else:
                    k -= 1


ThreeSum()
